### ISSUE TITLE
Bug: Off-by-One Token Error in Multiple Choice Decoding (Qwen3-VL & Qwen2.5-VL)

### ISSUE THREAD

**Issue by pulkitkumar95** [2U1/Qwen2-VL-Finetune#228]
I'm encountering an issue when fine-tuning on a multiple-choice question dataset using both Qwen3-VL and Qwen2.5-VL models. When decoding the predicted answer, it consistently appears to be off by one token compared to the target text, usually resulting in the loss of the first character (e.g., the option letter).

This suggests a potential issue in the decoding logic or tokenization handling for the fine-tuned models, possibly related to how the start-of-sequence or target text is processed during training/decoding.

---

### 📝 Examples

| Target Text | Predicted Text | Observation |
| :--- | :--- | :--- |
| C. right movement | . right movement | Missing 'C' |
| A. The bird is singing. | . The bird is singing. | Missing 'A' |
| D. 42 | . 42 | Missing 'D' |

The predicted text is always one token shorter and begins with the second token of the target.

This is the code that I used to check decoding:

```python
def check_decode(tokenizer, labels, logits, ignore_index=-100):
    if isinstance(logits, torch.Tensor):
        logits = logits.detach().cpu().numpy()
    if isinstance(labels, torch.Tensor):
        labels = labels.detach().cpu().numpy()

    predictions = np.argmax(logits, axis=-1)

    first_letter_results = []
    for target_ids, pred_ids in zip(labels, predictions):
        answer_mask = target_ids != ignore_index
        if not np.any(answer_mask):
            continue

        target_tokens = target_ids[answer_mask].tolist()
        pred_tokens = pred_ids[answer_mask].tolist()


        target_text = tokenizer.decode(target_tokens, skip_special_tokens=True).strip()
        pred_text = tokenizer.decode(pred_tokens, skip_special_tokens=True).strip()
        print('target_text:', target_text)
        print('pred_text:', pred_text)

    return
```

Do you have any idea why this might be happening?

**Comment by 2U1** [comment]
"Missing first character / Off by one token" pattern iswhat you get when you decode argmax(logits) without the causal shift.
1. logits[t] predicts input_ids[t+1]
2. so the prediction for the first answer token (the C in C.) lives at the last prompt position.
3. but your answer_mask selects positions where labels != -100 (answer positions), and then you take pred_ids[answer_mask] at those same positions → that corresponds to predicting the next answer token (so you start at . instead of C).


It’s your debug decode alignment.

You could fix your `check_decode` function by shifting like the model does for loss.
```
import numpy as np
import torch

def check_decode(tokenizer, labels, logits, ignore_index=-100):
    if isinstance(logits, torch.Tensor):
        logits = logits.detach().cpu().numpy()
    if isinstance(labels, torch.Tensor):
        labels = labels.detach().cpu().numpy()

    # logits: [B, T, V]
    preds = np.argmax(logits, axis=-1)  # [B, T]

    # Causal alignment: compare preds[t] with labels[t+1]
    shift_preds  = preds[:, :-1]
    shift_labels = labels[:, 1:]

    for target_ids, pred_ids in zip(shift_labels, shift_preds):
        mask = target_ids != ignore_index
        if not np.any(mask):
            continue

        target_tokens = target_ids[mask].tolist()
        pred_tokens   = pred_ids[mask].tolist()

        print("target_text:", tokenizer.decode(target_tokens, skip_special_tokens=True))
        print("pred_text  :", tokenizer.decode(pred_tokens,   skip_special_tokens=True))

```

For the inference, you could check like this

```
generated_ids = model.generate(**inputs, max_new_tokens=128)
generated_ids_trimmed = [
    out_ids[len(in_ids):] for in_ids, out_ids in zip(inputs.input_ids, generated_ids)
]
text = processor.batch_decode(generated_ids_trimmed, skip_special_tokens=True, clean_up_tokenization_spaces=False)
```
